src/Temporalio/Bridge/CustomSlotSupplier.cs

```diff
@@ -0,0 +1,179 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Temporalio.Bridge.Interop;
+
+namespace Temporalio.Bridge
+{
+    /// <summary>
+    /// Core wrapper for a user-defined custom slot supplier.
+    /// </summary>
+    internal class CustomSlotSupplier : NativeInvokeableClass<Interop.CustomSlotSupplierCallbacks>
+    {
+        private readonly ILogger logger;
+        private readonly Temporalio.Worker.Tuning.ICustomSlotSupplier userSupplier;
+        private readonly Dictionary<uint, Temporalio.Worker.Tuning.ISlotPermit> permits = new();
+        private uint permitId = 1;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CustomSlotSupplier" /> class.
+        /// </summary>
+        /// <param name="userSupplier">User's slot supplier implementation'.</param>
+        /// <param name="loggerFactory">Logger factory.</param>
+        internal unsafe CustomSlotSupplier(
+            Temporalio.Worker.Tuning.ICustomSlotSupplier userSupplier,
+            ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger<CustomSlotSupplier>();
+            this.userSupplier = userSupplier;
+
+            var interopCallbacks = new Interop.CustomSlotSupplierCallbacks
+            {
+                reserve = FunctionPointer<Interop.CustomReserveSlotCallback>(Reserve),
+                cancel_reserve = FunctionPointer<Interop.CustomCancelReserveCallback>(CancelReserve),
+                try_reserve = FunctionPointer<Interop.CustomTryReserveSlotCallback>(TryReserve),
+                mark_used = FunctionPointer<Interop.CustomMarkSlotUsedCallback>(MarkUsed),
+                release = FunctionPointer<Interop.CustomReleaseSlotCallback>(Release),
+            };
+
+            PinCallbackHolder(interopCallbacks);
+        }
+
+        private static void SetCancelTokenOnCtx(ref SlotReserveCtx ctx, CancellationTokenSource cancelTokenSrc)
+        {
+            unsafe
+            {
+                try
+                {
+                    var handle = GCHandle.Alloc(cancelTokenSrc);
+                    fixed (Interop.SlotReserveCtx* p = &ctx)
+                    {
+                        Interop.Methods.set_reserve_cancel_target(p, GCHandle.ToIntPtr(handle).ToPointer());
+                    }
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine($"Error setting cancel token on ctx: {e}");
+                    throw;
+                }
+            }
+        }
+
+        private unsafe void Reserve(Interop.SlotReserveCtx ctx, void* sender)
```

Comment:
```suggestion
        private unsafe void Reserve(Interop.SlotReserveCtx* ctx, void* sender)
```

I would have expected these all to be pointers so they aren't copied (so `*const SlotReserveCtx` on the Rust side). I think it's clearer that these short-lived Rust-owned things be passed by reference to Rust, so you don't need to re-capture a fixed pointer later.

Reply:
Yeah, I was thinking about doing that, makes sense.